### DISCUSSION TITLE
BluetoohLE - Best way to manually read Characteristic value on a given interval?

### DISCUSSION THREAD

**Discussion by wagenheimer** [shinyorg/shiny#1142] [Q&A] [answered]
I need to read some Characteristic values each given interval.

I do use Device.StartTimer but I am having problems writing in a Characteristic of this device, which takes too long.

I wanted to disable the readouts while I'm trying to write, but this way doesn't seem to be threadsafe and I can't set a variable to prevent readouts while writing.

What would be the correct way to do this?

**Comment by aritchie** [comment, marked as answer]
Android prevents a read/write at the same time (it has to).  iOS does not, but does its own internal synchronization.

```cs
Observable
    .Interval(TimeSpan.FromSeconds(30))
    .Where(_ => !this.IsWriting)
    .Select(_ => yourChar.Read())
    .Switch()
    .Subscribe(
        x => {}, // do whatever with the read value
        ex => {} // deal with error
   );
```

I'm not sure why you can't set a var when writing, but you need to for your requirement

**Comment by wagenheimer** [comment]
Hi! Thanks for the help!

But I'm facing the same problem with Observable.Interval that I was when using Timers.

```
IsWriting=true;
await ManagedPeripheral.Write(serviceUuid, characteristicUuid, valueToSend);
IsWriting=false;
```

For some reason, it still keeps reading the values even with IsBusy=true, and to Write a single bit takes several seconds (And I can see the UI still updating with new values while still awaiting for the writing. 

I'm thinking it's a ThreadSafe problem. Do you have any idea why this is occurring and the correct way to resolve this?

**Reply by aritchie** [reply]
No idea.  I gave the solution I use in a project.  You can always interlock the boolean if you have that big of a race condition.
